knack/arguments.py

```diff
@@ -143,6 +143,89 @@ def __enter__(self):
+
+            parent_class = self._get_parent_class(**kwargs)
+
+            class DeprecatedArgumentAction(parent_class):
+
+                def __call__(self, parser, namespace, values, option_string=None):
+                    if not hasattr(namespace, '_argument_deprecations'):
+                        setattr(namespace, '_argument_deprecations', [deprecate_info])
+                    else:
+                        namespace._argument_deprecations.append(deprecate_info)  # pylint: disable=protected-access
+                    try:
+                        super(DeprecatedArgumentAction, self).__call__(parser, namespace, values, option_string)
+                    except NotImplementedError:
+                        pass
+
+            return DeprecatedArgumentAction
+
+        def _handle_option_deprecation(deprecated_options):
+
+            if not isinstance(deprecated_options, list):
+                deprecated_options = [deprecated_options]
+
+            parent_class = self._get_parent_class(**kwargs)
+
+            class DeprecatedOptionAction(parent_class):
+
+                def __call__(self, parser, namespace, values, option_string=None):
+                    deprecated_opt = next((x for x in deprecated_options if option_string == x.target), None)
+                    if deprecated_opt:
+                        if not hasattr(namespace, '_argument_deprecations'):
+                            setattr(namespace, '_argument_deprecations', [deprecated_opt])
+                        else:
+                            namespace._argument_deprecations.append(deprecated_opt)  # pylint: disable=protected-access
+                    try:
+                        super(DeprecatedOptionAction, self).__call__(parser, namespace, values, option_string)
+                    except NotImplementedError:
+                        pass
+
+            return DeprecatedOptionAction
+
+        action = kwargs.get('action', None)
+
+        deprecate_info = kwargs.get('deprecate_info', None)
+        if deprecate_info:
+            deprecate_info.target = deprecate_info.target or argument_dest
+            action = _handle_argument_deprecation(deprecate_info)
+        deprecated_opts = [x for x in kwargs.get('options_list', []) if isinstance(x, Deprecated)]
+        if deprecated_opts:
+            action = _handle_option_deprecation(deprecated_opts)
+        return action
+
+    def deprecate(self, **kwargs):
+
+        def _get_deprecated_arg_message(self):
+            line1 = "{} '{}' has been deprecated and will be removed ".format(
+                self.object_type, self.target).capitalize()
+            if self.expiration:
+                line1 += "in version '{}'.".format(self.expiration)
+            else:
+                line1 += 'in a future release.'
+            lines = [line1]
+            if self.redirect:
+                lines.append("Use '{}' instead.".format(self.redirect))
+            return ' '.join(lines)
```

Comment:
why not just add to line1 if `self.redirect`?
lines will only have 1 or 2 elements.

Or, create the list at the beginning of this method, seems weird that you interchange between adding and then creating the list after the second line is added.